### PR TITLE
RN-899: Make the CheckboxList component generic

### PR DESCRIPTION
**Pull request by rohan-bes**
Making this component generic allows us to override the base `ListItem` type when using the `CheckboxList`.

Just to address this comment thread: https://github.com/beyondessential/tupaia/pull/4848/files#r1296553278

### CODE REVIEW

File: packages/ui-components/src/components/CheckboxList.tsx [hunk 1]
```diff
@@ -89,47 +89,45 @@ export interface ListItemProps extends Record<string, unknown> {
...
+
+  const intersection = (a: ListItem[], b: ListItem[]) => {
+    return a.filter(
+      item => b.findIndex(i => i[i.valueKey || 'code'] === item[item.valueKey || 'code']) !== -1,
+    );
+  };
+
+  const union = (a: ListItem[], b: ListItem[]) => {
+    return [...a, ...not(b, a)];
+  };
```
Comment (rohan-bes) [inline packages/ui-components/src/components/CheckboxList.tsx:117]: Kind of a bummer that we need to move these function in to the component. We could make them generic themselves, but I didn't quite feel like writing so much boilerplate 😅

File: packages/tupaia-web/src/api/mutations/useExportDashboard.tsx [hunk 2]
```diff
@@ -24,7 +24,7 @@ export const useExportDashboard = ({ onSuccess }: { onSuccess?: (data: Blob) =>
         selectedDashboardItems: selectedDashboardItems?.join(','),
       });
       // Auth cookies are saved against this domain. Pass this to server, so that when it pretends to be us, it can do the same.
-      const cookieDomain = new URL(import.meta.env.REACT_APP_TUPAIA_WEB_API_URL).hostname;
```
Comment (rohan-bes) [inline packages/tupaia-web/src/api/mutations/useExportDashboard.tsx:27]: Got caught out by not having this env var set during testing. So referenced the value in `api.ts`

File: packages/tupaia-web/src/features/Dashboard/ExportDashboard.tsx [hunk 3]
```diff
@@ -50,14 +50,18 @@ const ButtonGroup = styled.div`
 interface ExportDashboardProps {
   isOpen: boolean;
   onClose: () => void;
-  dashboardItems?: DashboardItem[];
+  dashboardItems: DashboardItem[];
```
Comment (acdunham) [inline packages/tupaia-web/src/features/Dashboard/ExportDashboard.tsx:53]: My only comment here would be when the dashboard items are loading they may be undefined

**Review by acdunham — APPROVED** [review]
Thanks for doing this! I had one question around if `dashboardItems` is undefined (eg not loaded yet) but other than that pre-approving

File: packages/tupaia-web/src/features/Dashboard/ExportDashboard.tsx (hunk 3 above)
Comment (rohan-bes) [reply packages/tupaia-web/src/features/Dashboard/ExportDashboard.tsx:53]: I added a default `dashboardItems = []` in the component args, so it should always have a value now. Seems like we were doing something similar before with defaulting to an empty list if the `dashboardItems` were `undefined` and I felt this was a little cleaner 👍

**Pull request merged**